src/main/java/dev/morling/onebrc/CalculateAverage_royvanrijn.java

```diff
@@ -74,29 +78,24 @@ private static Unsafe initUnsafe() {
     }
 
     public static void main(String[] args) throws Exception {
-
         // Calculate input segments.
         final int numberOfChunks = Runtime.getRuntime().availableProcessors();
         final long[] chunks = getSegments(numberOfChunks);
 
-        final List<Entry[]> repositories = IntStream.range(0, chunks.length - 1)
+        final Map<String, Entry> measurements = HashMap.newHashMap(1 << 10);
+        IntStream.range(0, chunks.length - 1)
                 .mapToObj(chunkIndex -> processMemoryArea(chunks[chunkIndex], chunks[chunkIndex + 1]))
                 .parallel()
```

Comment:
Have you tried manual parallelization with an array of `Thread`s?

Reply:
I'm not familiar with all the details of @royvanrijn's implementation, but I imagine that if you have more chunks than (system) threads available (mine implementation may, if the chunks are too large to all fit within a `mmap`) you might have parked threads. I _assume_ `parallel` uses work-stealing when a thread is done. I'm not sure how java manages the virtual threads, so this entire comment may be irrelevant

Reply:
It doesn't, I've tried submitting parts and starting to process once the CompletableFuture is done; but all of that adds overhead more than it solves a problem.

The parallel() can't really work-steal, the ranges are set by the chunk array.

Reply:
what if you use counted completers, your main thread will not split to N segments but start walking trough the data to identify separator, each time it identifies other workers can steal and process the entry... once main is done identifying all of the entries it joins and consumes whatever is left.

Basically splitting to N chunks (which are not really the same size +- 2 entires) means your slowest worker is the number you get. Since in theory you are doing the walk anyway and identifying the separators you are not doing any extra work but using true work steal... 

You can probably improve the speed of seeking the separators if yu know how many bytes you can confidently skip after last separator... the name and whatever boilerplate is in the file will give you some offset that you can skip reliably. I'm not sure if names have minimum length etc..

Reply:
though i will assume the volatile write that happens to get things synchronised will cost you more than you benefit from perfectly distributing work...